Add OCP-43493 into service.feature

jechen0648 opened the pull request:
@weliang1 @anuragthehatter  @zhaozhanqi  Please review this PR, here is the runner log:
https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/Runner-v3-smoke/3750/console

zhaozhanqi commented on features/networking/service.feature:
```diff
@@ -633,3 +633,82 @@ Feature: Service related networking scenarios
+      | exec_command | /host                        |
+      | exec_command | curl                         |
+      | exec_command | 1.1.1.1:27017                |
+    Then the step should succeed
```
we can use 
```
    Given I use the "<%= cb.nodes[0].name %>" node
    And I run commands on the host:
      | curl 1.1.1.1:27017 |
```

jechen0648 replied:
Thanks, I have used this method.  But for next step of patch changing externalIP for svc, I had to add the following to re-enforce the "oc patch" command is executed with right project.  I found after curil 1.1.1.1:27017 using the method you suggested, it has left my project.
     Given I switch to cluster admin pseudo user
     And I use the "<%= cb.proj_name %>" project

zhaozhanqi replied:
sorry, I forgot to mentioned the `--connect-timeout 5` when `curl`

anuragthehatter commented:
cc @openshift/team-sdn-qe

jechen0648 commented:
@weliang1 @zhaozhanqi @anuragthehatter  can you review again?  here is the runner log with updated script: https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/Runner-v3-smoke/3757/console

@openshift/team-sdn-qe

weliang1 commented on features/networking/service.feature:
```diff
@@ -633,3 +633,74 @@ Feature: Service related networking scenarios
+    And the output should contain "Hello OpenShift!"
+
+    # change externalIP for the svc
+    Given I switch to cluster admin pseudo user 
```
The normal user will update externalIP for the svc not cluster admin, cluster admin only define the policy not svc. The step in line 690 should be use the way in line 663 as "Given I switch to the first user"

jechen0648 replied:
I had tried, if I use "Given I switch to the first user" with "And I use the "<%= cb.proj_name %>" project", I got "service-unsecure not found" error, my script stuck at trying to find "service-unsecure" to change its externalIP, but if I use "Given I switch to cluster admin pseudo user " with "And I use the "<%= cb.proj_name %>" project", it can find the service and patch change its externalIP address.  So I still believe I should use cluster admin

weliang1 replied:
You can use cluster admin in the script, but that is not the user case and defeat the testing purpose. Could you try to manual test by switching between admin and test user? It should work.

jechen0648 replied:
manually, yes, I was able to patch change externalIP address for the svc.  But with script using first user, it can not find the svc.  I am asking Peter to help looking into it

weliang1 commented on features/networking/service.feature:
```diff
@@ -633,3 +633,74 @@ Feature: Service related networking scenarios
+    # change externalIP for the svc
+    Given I switch to cluster admin pseudo user 
+    And I use the "<%= cb.proj_name %>" project
+    And I wait up to 300 seconds for the steps to pass:
```
I do not think you need 300 seconds here, 500 seconds in other cases is because change the new externalIP policy in the CNO will take that long to take effect. You case is update svc not externalIP policy

jechen0648 commented:
@weliang1 @anuragthehatter @zhaozhanqi   @openshift/team-sdn-qe 
Made changes to Weibin's review comments:
1) reduce waiting time from 300s to 60s when changing externalIP for svc;  
2) use "first user" to change externalIP for svc
Here is the latest runner log:  https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/Runner-v3-smoke/3765/console

weliang1 commented on features/networking/service.feature:
```diff
@@ -633,3 +633,75 @@ Feature: Service related networking scenarios
+    And the pod named "externalip-pod" becomes ready
+ 
+    # Curl externalIP:portnumber should pass
+    Given I switch to cluster admin pseudo user
```
Actually you do not need switch to admin, line 702-704 just works. Seems "Given I use the "<%= cb.nodes[0].name %>" node" will automatically use admin mode.
I tired and it just worked: http://pastebin.test.redhat.com/1012092

jechen0648 commented:
sure, I removed the "switching to cluster pesudo admin user", here is the latest runner log: https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/Runner-v3-smoke/3766/console
Please review it again, thanks!

weliang1 commented:
/lgtm

asood-rh commented on features/networking/service.feature:
```diff
@@ -633,3 +633,73 @@ Feature: Service related networking scenarios
+    # Add externalIP policy to CNO
+    Given as admin I successfully merge patch resource "networks.config.openshift.io/cluster" with:
+      | {"spec":{"externalIP":{"policy":{"allowedCIDRs":["1.1.1.0/24"]}}}} |
+ 
```
Is there a reason the network operator is not being checked after the patch request?

zhaozhanqi replied:
@asood-rh I guess the following steps 500 seconds should be able to make sure it works. 
```
    And I wait up to 500 seconds for the steps to pass:
    """
    When I run oc create over "externalip_service1.json" replacing paths:
      | ["spec"]["externalIPs"][0] | 1.1.1.1 |
    Then the step should succeed
    """
```

weliang1 replied:
@asood-rh,  @zhaozhanqi is correct. Service can not be applied and will log the error until network operator become available again. For externalIP testing,  it normally take about 500 seconds.

jechen0648 commented:
Made more changes:
1) added check on network operator, since I check network operator at front, I reduced the waiting time at creating service
2) added --connect-timeout 5 when curl

Here is latest runner log:https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/Runner-v3-smoke/3783/console

Please review again, @asood-rh @weliang1 @zhaozhanqi @openshift/team-sdn-qe

weliang1 commented on features/networking/service.feature:
```diff
@@ -633,3 +633,78 @@ Feature: Service related networking scenarios
+    Given I switch to the first user
+    And I use the "<%= cb.proj_name %>" project
+    Given I obtain test data file "networking/externalip_service1.json"
+    And I wait up to 300 seconds for the steps to pass:
```
If you check network operator status before, then you do not need wait for 300 seconds just to create a service.

jechen0648 replied:
I tried to reduce to 60s, I had one failure at local run, I really do not have feeling how much I should wait, that is why I added back to 300s, but I do not think it waits full 300s in reality, if just periodically polls, and if it pass, it would move on before reaching 300s.

jechen0648 replied:
@weliang1 not going to check network operator any more, as it is not reliable, I will just wait 500s at creating externalip-service

jechen0648 commented:
Checking network operator status does not seem to be reliable, I am copy pasting portion of my local running log to show that although it passed network operator check (I personally think it is false positive), network operator may take some time to react on the change.    So I am going abardon the checking network operator idea.  
waiting for operation up to 500 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
    And I wait up to 500 seconds for the steps to pass:                                             # features/step_definitions/meta_steps.rb:33
      """
        Given the status of condition "Progressing" for network operator is :False
        And the status of condition "Available" for network operator is :True
      """
      [22:59:03] INFO> Shell Commands: oc get clusteroperators network -o jsonpath\=\{.status.conditions\[\?\(.type\ \=\=\ \"Progressing\"\)\].status\} --kubeconfig=/home/jechen/workdir/jechen-jechenx/ocp4_admin.kubeconfig
      False
      [22:59:04] INFO> Exit Status: 0
      [22:59:04] INFO> Shell Commands: oc get clusteroperators network -o jsonpath\=\{.status.conditions\[\?\(.type\ \=\=\ \"Available\"\)\].status\} --kubeconfig=/home/jechen/workdir/jechen-jechenx/ocp4_admin.kubeconfig
      True
      [22:59:05] INFO> Exit Status: 0
    # Clean-up required to erase above externalIP policy after testing done
    Given I register clean-up steps:                                                                # features/step_definitions/meta_steps.rb:9
      """
      Given as admin I successfully merge patch resource "networks.config.openshift.io/cluster" with:
        | {"spec":{"externalIP":{"policy":{"allowedCIDRs": null}}}} |
      """
    # Create a svc with externalIP
    Given I switch to the first user                                                                # features/step_definitions/user.rb:2
waiting for operation up to 3600 seconds..
    And I use the "<%= cb.proj_name %>" project                                                     # features/step_definitions/project.rb:110
      [22:59:05] INFO> Shell Commands: oc project 8vbhq --kubeconfig=/home/jechen/workdir/jechen-jechenx/ocp4_testuser-4.kubeconfig
      Already on project "8vbhq" on server "https://api.jechen-1201b.qe.gcp.devcluster.openshift.com:6443".
      
      [22:59:06] INFO> Exit Status: 0
cp /home/jechen/automation-work/verification-tests/testdata/networking/externalip_service1.json externalip_service1.json
    Given I obtain test data file "networking/externalip_service1.json"                             # features/step_definitions/file.rb:1
waiting for operation up to 120 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
waiting for operation up to 3600 seconds..
    And I wait up to 120 seconds for the steps to pass:                                             # features/step_definitions/meta_steps.rb:33
      """
      When I run oc create over "externalip_service1.json" replacing paths:
        | ["spec"]["externalIPs"][0] | 1.1.1.1 |
      Then the step should succeed
      """
      [22:59:06] INFO> {"kind":"Service","apiVersion":"v1","metadata":{"name":"service-unsecure","labels":{"name":"service-unsecure"}},"spec":{"ports":[{"name":"http","protocol":"TCP","port":27017,"targetPort":8080}],"externalIPs":["1.1.1.1"],"selector":{"name":"externalip-pod"}}}
      [22:59:06] INFO> Shell Commands: oc create -f - --kubeconfig=/home/jechen/workdir/jechen-jechenx/ocp4_testuser-4.kubeconfig
      
      STDERR:
      Error from server (Forbidden): error when creating "STDIN": services "service-unsecure" is forbidden: spec.externalIPs: Forbidden: externalIPs have been disabled
      
      [22:59:07] INFO> Exit Status: 1
      [23:01:06] INFO> last 4 messages repeated 59 times
      the step failed (RuntimeError)
      /home/jechen/automation-work/verification-tests/features/step_definitions/common.rb:6:in `/^the step should( not)? (succeed|fail)$/'
      /home/jechen/automation-work/verification-tests/features/step_definitions/transform.rb:33:in `call'
      /home/jechen/automation-work/verification-tests/features/step_definitions/transform.rb:33:in `block in singleton class'
      /home/jechen/automation-work/verification-tests/features/step_definitions/meta_steps.rb:48:in `block (2 levels) in <top (required)>'
      /home/jechen/automation-work/verification-tests/lib/base_helper.rb:155:in `wait_for'
      /home/jechen/automation-work/verification-tests/features/step_definitions/meta_steps.rb:43:in `/^I wait(?: up to ([0-9]+|<%=.+?%>) seconds)? for the steps to pass:$/'
      features/networking/service.feature:671:in `I wait up to 120 seconds for the steps to pass:'
    # Create a pod

jechen0648 commented:
So I am abandoning checking network operator idea.  I will just wait 500s at creating externalip_service like everyone else did.   
Here is latest long: https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/Runner-v3-smoke/3784/console
Please review again, thanks!  @asood-rh @weliang1 @zhaozhanqi @openshift/team-sdn-qe

zhaozhanqi commented:
/lgtm

weliang1 commented:
/lgtm

zhaozhanqi commented:
/approve
